Readme.md

```diff
@@ -0,0 +1,369 @@
+
+Notes:
+- Status could be updated by Job task using `Arc<RwLock<JobOutcome>>`. Alternative approach I've considered was to use `Arc<AtomicU16>`, but that would require magic numbers. Given that value is updated once when the job is complete or terminated I believe `RwLock` is appropriate synchronization tool to use.  
+
+<details>
+<summary> Job status mermaid </summary>
+
+```txt
+sequenceDiagram
+actor CA as Client A
+participant S as Job Runner Service 
+participant CRA as Controller for Client A
+
+Note over CA,CRA: Job with id JobId for Client A was started earlier
+CA->>S: Status(JobId)
+S->>CRA: Status for JobId pls
+alt Job still running
+CRA->>S: Id exists, no status, must be running
+S->>CA: running
+else Job exited
+CRA->>S: Id exists, has exit_code
+S->>CA: exit_code
+else Client kills Job
+CRA->>S: Id exists, has signal
+S->>CA: signal
+end 
+```
+</details>
+
+### Output
+
+[![](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5hY3RvciBDQSBhcyBDbGllbnQgQVxucGFydGljaXBhbnQgUyBhcyBKb2IgUnVubmVyIFNlcnZpY2UgXG5wYXJ0aWNpcGFudCBDUkEgYXMgQ29udHJvbGxlciBmb3IgQ2xpZW50IEFcbnBhcnRpY2lwYW50IEogYXMgSm9iXG5wYXJ0aWNpcGFudCBGUyBhcyBGaWxlIFN5c3RlbVxuXG5Ob3RlIG92ZXIgQ0EsRlM6IEpvYiB3aXRoIGlkIEpvYklkIGZvciBDbGllbnQgQSB3YXMgc3RhcnRlZCBlYXJsaWVyXG5KLSlGUzogU3RyZWFtcyBsb2dzXG5DQS0-PlM6IE91dHB1dChKb2JJZClcblMtPj5DUkE6IE91dHB1dCBmb3IgSm9iSWQgcGxzXG5DUkEtPj5GUzogT3BlbiBMb2cgZmlsZXMgZm9yIEpvYklkXG5GUy0-PkNSQTogSGVyZSBpcyB0aGUgaGFuZGxlXG5DUkEtPj5TOiBTdHJlYW0gY29udGVudHMgYmFja1xuUy0-PkNBOiBIZXJlIGFyZSB5b3VyIGxvZ3MiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)](https://mermaid.live/edit#eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5hY3RvciBDQSBhcyBDbGllbnQgQVxucGFydGljaXBhbnQgUyBhcyBKb2IgUnVubmVyIFNlcnZpY2UgXG5wYXJ0aWNpcGFudCBDUkEgYXMgQ29udHJvbGxlciBmb3IgQ2xpZW50IEFcbnBhcnRpY2lwYW50IEogYXMgSm9iXG5wYXJ0aWNpcGFudCBGUyBhcyBGaWxlIFN5c3RlbVxuXG5Ob3RlIG92ZXIgQ0EsRlM6IEpvYiB3aXRoIGlkIEpvYklkIGZvciBDbGllbnQgQSB3YXMgc3RhcnRlZCBlYXJsaWVyXG5KLSlGUzogU3RyZWFtcyBsb2dzXG5DQS0-PlM6IE91dHB1dChKb2JJZClcblMtPj5DUkE6IE91dHB1dCBmb3IgSm9iSWQgcGxzXG5DUkEtPj5GUzogT3BlbiBMb2cgZmlsZXMgZm9yIEpvYklkXG5GUy0-PkNSQTogSGVyZSBpcyB0aGUgaGFuZGxlXG5DUkEtPj5TOiBTdHJlYW0gY29udGVudHMgYmFja1xuUy0-PkNBOiBIZXJlIGFyZSB5b3VyIGxvZ3MiLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)
+
+Notes:
+- Considered keeping logs in-memory, but I believe this approach has couple of drawbacks. Reads and writes need to be synchronized somehow, so our options are either introducing Mutex, which may become fairly hot (depends on the volume of logs), or making Job own the output of the child process instead of streaming to file. This consequently means that all job output is now in main process memory, regardless of whether it will ever be read or not.  
```

Comment:
How are clients going to get the new output produced by the job in real-time?

Reply:
Good callout, added a note on `inotify`